providers/implementations/rands/seeding/rand_unix.c

```diff
@@ -717,6 +718,12 @@ size_t ossl_pool_acquire_entropy(RAND_POOL *pool)
         return entropy_available;
 #   endif
 
+#   if defined(OPENSSL_RAND_SEED_JITTER)
+    entropy_available = ossl_prov_acquire_entropy_from_jitter(pool);
+    if (entropy_available > 0)
+        return entropy_available;
+#   endif
```

Comment:
I'm not convinced that the OS source should include jitter.

Reply:
I don't believe this is of any interest in the generic provide.

The kernel provided entropy is typically of better quality, more performant and more portable.

I am weary of changing how the three default entropy drgbs look like.

I am weary of users using built-in / always available seed-src if it doesn't provide certified entropy.

I am very cautious of introducing jitter as top level supported RNG. It's not performant enough to be exposed like that. See comments in aws-lc.

I can only commit to supporting this code to be used for seeding rngs only, without direct use/exposure per se.

Certification will be done on the static libjitterentropy.a standalone without openssl double conditioning.

And slightly related, not too sure why chaining is done from seed to primary to then (public,private). My unbenchmarked hunch is that in high performance scenarios locking of primary likely has more overhead than seeding public/private directly from getrandom. Especially on all stable Linux kernels post-random refactor patches. I don't have insight on non-linux OS.

If jitter is made a top level RNG, would primary/private/public DRGBs set it as their parent in only fips, only default, build/runtime/either configurable?

Reply:
I am feeling very conservative about this. And would rather first introduce it as Linux only, inside seed-src, and likely exclusive, and like under fips only.

If that works out fine, I am happy to help expand it to other arches, default provider, and/or graduating it to a top level standalone rng. But not straightaway. It might be missused, resulting in poor performance and regarded as a regression.

Reply:
The project's extremely conservative approach to changes makes putting in, as you suggest, and later changing it somewhat problematic: it is best to aim for the final solution directly to avoid creating legacy baggage that must be supported going forwards.  I still think a separate top level seed source is the approach here.

With jitter as a top level entropy source, it suffices to set it as the parent of the primary DRBG or even as the primary rand source.  These can be done via configuration and/or programmatically.  The infrastructure is already in place to do this.  What isn't fully in place is the conditional RNG tests that NIST mandates when the FIPS provider receives entropy.  They are done when collecting via the callback to libcrypto but aren't done otherwise.  I'm working on this currently.  It's either really nasty or requires a new provider API (which means more legacy ...).

From a FIPS perspective, validating libjitterentropy.a separately introduces another problem: to move data between validated FIPS modules, a secure channel must be used.  This means ensuring integrity of the data transferred.  This isn't a problem for a whole of device validation as you do but would be a consideration for the project.  I somewhat doubt the project would validate jitter as an entropy source.  It's been discussed and we support too many platforms where it's not going to work well (or at all).

The primary/public/private setup was designed to reduce the amount of entropy required.  For hosts with slow RNGs this matters.  On modern kernels with decent processors, it's not a concern of course but the project supports lots of architectures.  A surprising number of small embedded boxes run OpenSSL and these tend to be terrible at collecting entropy.

Reply:
> The project's extremely conservative approach to changes makes putting in, as you suggest, and later changing it somewhat problematic: it is best to aim for the final solution directly to avoid creating legacy baggage that must be supported going forwards. I still think a separate top level seed source is the approach here.
> 

Also true.

> With jitter as a top level entropy source, it suffices to set it as the parent of the primary DRBG or even as the primary rand source. These can be done via configuration and/or programmatically. The infrastructure is already in place to do this. What isn't fully in place is the conditional RNG tests that NIST mandates when the FIPS provider receives entropy. They are done when collecting via the callback to libcrypto but aren't done otherwise. I'm working on this currently. It's either really nasty or requires a new provider API (which means more legacy ...).
> 

But statically linked jitterentropy.a inside fips.so is not external entropy source, but an internal one. I will check what RNG tests are needed there.

> From a FIPS perspective, validating libjitterentropy.a separately introduces another problem: to move data between validated FIPS modules, a secure channel must be used. This means ensuring integrity of the data transferred.

When statically linked it is part of the validated FIPS module, because it is just an entropy source with ESV cert. This is no different to how kernel entropy sources with jitter are validate, which are considered inside the boundary and do not have any sort of secure channel.

This is in part why i'm not making jitter.so as a separate provider; nor linking jitter entropy as a shared library.

> This isn't a problem for a whole of device validation as you do but would be a consideration for the project.

I am not doing whole of device validation.

> I somewhat doubt the project would validate jitter as an entropy source. It's been discussed and we support too many platforms where it's not going to work well (or at all).
> 

Indeed that is not suitable for OpenSSL Project to validate. I see at least three vendors that are using jitterentropy with OpenSSL and they do seem to validate both as "openssl jitter" and as pure "jitter" entropy sources. It is mostly for those that already have or planning to get ESV cert for jitter and use it with OpenSSL in addition to other places.

> The primary/public/private setup was designed to reduce the amount of entropy required. For hosts with slow RNGs this matters. On modern kernels with decent processors, it's not a concern of course but the project supports lots of architectures. A surprising number of small embedded boxes run OpenSSL and these tend to be terrible at collecting entropy.

Yeap. And further kicker is that general purpose distributions are being executed on edge devices without clocks too.

Reply:
Today seed-src has lots of compile time options of types of seeds it supports. Some of them are incomplete, some of them are obsolete, and some of them are excluded from FIPS. A cleanup would be nice.

For FIPS, on Linux, on server-grade hardware it would be ideal to have a mix of getrandom/jitter => specifically, check if kernel is in FIPS mode, use getrandom; if kernel is not in FIPS mode fallback to jitter entropy. And ideally without chaining, when there is enough system entropy - as that avoids need for additional ESV certs.

Whilst one can compile such behaviour for default provider, and smaller devices it might just brick CPU harder rather than blocking on entropy. For those systems it would feel more appropriate to run jitterentropy-rngd or similar if it is determined that CPU is good enough.

Reply:
I'll try to explain better.

The way the FIPS provider is set up is that the primary DRBG calls out to libcrypto for it's entropy.  It does this via the CRNG testing: https://github.com/openssl/openssl/blob/b9e084f139c53ce133e66aba2f523c680141c0e6/providers/implementations/rands/drbg.c#L200-L204

The CRNG tests, in turn, call back to libcrypto for entropy:
https://github.com/openssl/openssl/blob/b9e084f139c53ce133e66aba2f523c680141c0e6/providers/implementations/rands/crngt.c#L35-L54
via
https://github.com/openssl/openssl/blob/b9e084f139c53ce133e66aba2f523c680141c0e6/providers/common/provider_seeding.c#L82-L92

The FIPS provider is consequently _only_ able to get entropy from libcrypto.  Due to the project's conservative nature about compatibility these callbacks have to remain and be supported.  The FIPS provider's conservatism is even worse: 3.x FIPS providers are guaranteed to work with all 3.y versions of OpenSSL & quite possibly will need to work with all 4.z versions too.  We know that currently, there are users who supply entropy to the FIPS provider via these callbacks and, consequently, we cannot break them.  The knowledge about what these callbacks are doing is simply not available to the FIPS provider, so we cannot detect what's happening and bypass them to use an internal entropy source directly.  At least without introducing something truly nasty.

Changes to SEED_SRC, as you've done here, have _zero_ impact on the FIPS provider because it is _outside_ the boundary.  Bringing SEED_SRC inside the boundary would require validation of all of the possible sources it contains, which simply isn't practical.  As this stands, it is of zero use for OpenSSL's FIPS validations.

Adding a new JITTER_SRC RNG would permit it to be inside the FIPS provider.  This would then be set as the entropy source for libcrpyto.  When the FIPS provider requires entropy, it does it's callback dance and ends up transferring the resulting data from itself to itself which could be justified as compliant without the secure channel stupidities.

If you were adding jitter as a source for non-FIPS use, the approach you've used here would be fine.  Since you're adding it for FIPS purposes, it isn't.  It **needs** to be its own top level entropy source.

Reply:
> The knowledge about what these callbacks are doing is simply not available to the FIPS provider, so we cannot detect what's happening and bypass them to use an internal entropy source directly.

> Changes to SEED_SRC, as you've done here, have zero impact on the FIPS provider because it is outside the boundary.

> Adding a new JITTER_SRC RNG would permit it to be inside the FIPS provider. This would then be set as the entropy source for libcrpyto. When the FIPS provider requires entropy, it does it's callback dance and ends up transferring the resulting data from itself to itself which could be justified as compliant without the secure channel stupidities.

I am still slightly confused by all of the above statements, because the first statement is what we actually try to sort of achieve. And I'm not sure what the difference is between "jitter_src_rng" and "seed_src(compiled to only have jitter as source seed)". Nor how a jitter_src rng can be brought into the boundary, given that https://github.com/openssl/openssl/pull/13226 intentially moved it out of the fips provider.

When fips provider is configured to only build with jitter entropy, why wouldn't the call 
 to ossl_crngt_get_entropy() be effectively replaced by calls to internal jitter entropy, without ever leaving the provider call paths? That way, together with all this existing code, it will be possible to achieve jitter-only rand inputs (for both libcrypto non-fips usage, and/or inside fips provider).

Reply:
@paulidale @t8m despite not being my intentions, I sort of like this implementation as is =))))))

When jitterentropy-library.la is a certified entropy source, and when libcrypto.so is configured to only use jitterentropy seed source (as in this pr / github workflow), existing FIPS providers (ie. 3.0.9) get access to userspace certified entropy source, without need to change them (tbd - cause it is not clear to me if entropy source requires tamper checks (hmac) and/or secure channel communication, when all of this is in process)

Separately, now looking into how I could patch a fips provider to not use the entropy callback and call into a statically linked jitter entropy source. Such that one can get compatibility the other way around - a jitter-using-fips-provider doesn't use callback to parent openssl for entropy.

Reply:
and i will still try to create a stand alone seed source, just to have all options available to us, and to really understand what is going on here.

Reply:
Sorry for the delay, I had to think about this.

Starting with the later question: I think that a stand alone seed source (from the FIPS provider) connected directly to a DRBG (also from the FIPS provider) doesn't need the secure channel stuff.  Data is going from the FIPS boundary directly to itself -- i.e. not passing through non FIPS areas.  In the unlikely event of the code flow outside the boundary being questioned, I'm pretty sure that it could be successfully argued as not being relevant -- the actual transfer of data is being done exclusively by code inside the boundary.

As for patching jitter directly into the FIPS provider, I think what you suggest would be workable for the FIPS provider but not for libcrypto.  libcrypto will always attempt to load the seed source which gives the primary DRBG a parent and this will be preferred in place of following the code paths you mention.  Thus, the FIPS providers internal uses of RAND_bytes() would be good, the external uses wouldn't be.

Reply:
" RAND_bytes() ... the external uses wouldn't be" not sure i even know how to use deprecated API as an external user. :disappointed: but will trace it. My naive expectation was that EVP usage would be fine, and that's all that matters.

Reply:
RAND_bytes() is by no means deprecated.

Reply:
I think the presence of the `ossl_prov_get_entropy()` call in the `get_entropy()` function in the `drbg.c` file is quite confusing.  I don't think OpenSSL can ever reach that line because we always set the primary DRBG's parent in libcrypto.  It's basically there for legacy purposes only.

I've raised #24745 to add a clarifying comment.

Reply:
> RAND_bytes() is by no means deprecated.

and never will be…

Reply:
Going back to these statements:

> Changes to SEED_SRC, as you've done here, have zero impact on the FIPS provider because it is outside the boundary. 

When libcrypto/libssl compiled with this patch set, and seed-source configured to jitter; any fips-provider will then use certified entropy source which as far as I can tell need not to be inside the openssl-fips-provider boundary. As in this patchset enables using 3.0.9 fips provider with jitter entropy source.

Looking at adding jitter-seed-src entropy provider which can be brought into the fips provider will get libcrypto & fips-provider to mostly use it, but not in full. If one doesn't patch seed-src (like I do in this PR), then `ossl_rand_get_entropy()` will remain a public API which would get OS entropy (i.e. getrandom syscall) instead of jitter-seed-src.

Sort of perplexed as to why ossl_rand_get_entropy() is a public API, instead of only exposing ossl_rand_get_user_entropy() which always goes via provider framework and would allow overriding entropy source.

I see benefit of splitting out seed-src into individual ones seed-src-getrandom seed-src-egd seed-src-rdcpu seed-src-jitter and so on.

Or maybe runtime config should be implemented for seed-src to allow toggling on/off getrandom / egd / rdcpu / jitter instead of having it compile time determined the priority of those.

Reply:
> Sort of perplexed as to why ossl_rand_get_entropy() is a public API, instead of only exposing ossl_rand_get_user_entropy() which always goes via provider framework and would allow overriding entropy source.

This is due to our supporting legacy code.  `ossl_rand_get_entropy()` was the first version of the call and it did something incorrect.  This necessitated `ossl_rand_get_user_entropy()` being added.  We cannot remove the former unless we go through the formal deprecation process (deprecated in a major release and removed in the following major release).